src/main/scala/io/iohk/ethereum/utils/RLP.scala

```diff
@@ -0,0 +1,276 @@
+        case p if p < OFFSET_SHORT_ITEM => (new RLPValue {
+          override def toBytes: Array[Byte] = Array(data(pos))
+        }, pos + 1)
+        case p if p <= OFFSET_LONG_ITEM =>
+          val length = p - OFFSET_SHORT_ITEM
+          val res: Array[Byte] = new Array[Byte](length)
+          Array.copy(data, pos + 1, res, 0, length)
+          (new RLPValue {
+            override def toBytes: Array[Byte] = res
+          }, pos + length + 1)
+        case p if p < OFFSET_SHORT_LIST =>
+          val lengthOfLength = p - OFFSET_LONG_ITEM
+          val lengthBytes = new Array[Byte](lengthOfLength)
+          Array.copy(data, pos + 1, lengthBytes, 0, lengthOfLength)
+          val length = decodeInt(lengthBytes)
+          val beginPos = pos + lengthOfLength + 1
+          val res = new Array[Byte](length)
+          Array.copy(data, beginPos, res, 0, length)
+          (new RLPValue {
+            override def toBytes: Array[Byte] = res
+          }, beginPos + length)
+        case p if p <= OFFSET_LONG_LIST =>
+          val length = p - OFFSET_SHORT_LIST
+          val (listDecoded, endPos) = decodeListRecursive(data, pos + 1, length, Queue()).get
+          (new RLPList {
+            override def items = listDecoded
+          }, endPos)
+        case p =>
+          val lengthOfLength = p - OFFSET_LONG_LIST
+          val lengthBytes = new Array[Byte](lengthOfLength)
+          Array.copy(data, pos + 1, lengthBytes, 0, lengthOfLength)
+          val length = decodeInt(lengthBytes)
+          val (listDecoded, endPos) = decodeListRecursive(data, pos + lengthOfLength + 1, length, Queue()).get
+          (new RLPList {
+            override def items = listDecoded
+          }, endPos)
+      }
+    }
+  }
+
+  @tailrec
+  private def decodeListRecursive(data: Array[Byte], pos: Int, length: Int,
+                                  acum: Queue[RLPEncodeable]): Try[(Queue[RLPEncodeable], Int)] = {
+    if (length == 0) Success((acum, pos))
+    else {
+      val maybeDecoded = decodeWithPos(data, pos)
+      maybeDecoded match {
+        case Success((decoded, decodedEnd)) => decodeListRecursive(data, decodedEnd, length - (decodedEnd - pos), acum :+ decoded)
+        case Failure(e) => Failure(e)
+      }
+    }
+  }
+}
+
+sealed trait RLPEncodeable
+
+trait RLPList extends RLPEncodeable {
+  def items: Seq[RLPEncodeable]
+
+  def sameElements(other: RLPList)(): Boolean = items == other.items
+
+  override def equals(other: Any): Boolean = other match {
+    case other: RLPList => hashCode() == other.hashCode()
+    case _ => false
+  }
+
+  override def hashCode(): Int = items.hashCode()
+}
+
+object RLPList {
+  def apply(enc: RLPEncodeable*): RLPList = new RLPList {
+    override def items: Seq[RLPEncodeable] = enc.toSeq
+  }
+
+  def apply[E](seq: Seq[E])(implicit convert: E => RLPEncodeable): RLPList = new RLPList {
+    override def items: Seq[RLPEncodeable] = seq.map(convert)
+  }
+}
+
+trait RLPValue extends RLPEncodeable {
```

Comment:
@LukasGasior1 let me see. I think it was left after a refactor and might be changed to case classes as you say.